### PR TITLE
[CARBONDATA-3506]Fix alter table failures on parition table with hive.metastore.disallow.incompatible.col.type.changes as true

### PR DESCRIPTION
**Pull request by akashrn5**
### Problem:
1. In case of spark2.2 and above and , when we call alterExternalCatalogForTableWithUpdatedSchema to update the new schema to external catalog in case of add column, spark gets the catalog table and then it itself adds the partition columns if the table is partition table for all the new data schema sent by carbon, so there will be duplicate partition columns, so validation fails in hive
2. When the table has only two columns and one of them is partition column, then dropping non partition column is invalid because, if we allow it is like table with all columns as partition columns. So with the above property as true, drop column will fail to update the hive metastore.
3. in spark2.2 and above if the datatype change is done on partition column, with the above property as true, it also fails, as we are not sending partition column for schema alter in hive

### Solution:
1. when sending the new schema to spark to update in catalog, do not send the partition columns in case of spark2.2 and above, as spark will take care of adding parition columns to new schema sent by us.
2. In the above scenario of drop, do not allow drop column, if after dropping the specific column, if table has only partition columns.
3. Block the operation on datatype change on partition column on spark2.2 and above.

Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [x] Any interfaces changed?
 NA
 - [x] Any backward compatibility impacted?
 NA
 - [x] Document update required?
NA
 - [x] Testing done
Added UT and tested in cluster.
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
       
 - [x] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA. 
NA

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/368/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/368/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/368/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/371/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/371/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/373/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/373/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/371/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/373/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/374/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/374/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/374/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/375/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/375/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/375/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/376/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/376/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/376/

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/standardpartition/StandardPartitionGlobalSortTestCase.scala [hunk 1]
```diff
@@ -1087,5 +1088,6 @@ class StandardPartitionGlobalSortTestCase extends QueryTest with BeforeAndAfterA
     sql("drop table if exists partitiondatadelete")
     sql("drop table if exists comp_dt2")
     sql("drop table if exists partitionNoColumn")
+    sql("drop table if exists part_alter")
```
Comment (ravipesala) [inline integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/standardpartition/StandardPartitionGlobalSortTestCase.scala]: remove if not needed

**Comment by ravipesala** [comment]
LGTM apart from a small suggestion.

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/standardpartition/StandardPartitionGlobalSortTestCase.scala (hunk 1 above)
Comment (akashrn5) [reply integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/standardpartition/StandardPartitionGlobalSortTestCase.scala]: Done

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/381/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/381/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/381/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/409/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/409/

**Comment by akashrn5** [comment]
@ravipesala added a commit to block alter datatype change on partition column, please help to review

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/408/

**Comment by kumarvishal09** [comment]
LGTM

**Comment by kumarvishal09** [comment]
LGTM

**Comment by kumarvishal09** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/431/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/431/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/428/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/472/

**Comment by kumarvishal09** [comment]
LGTM

**Pull request closed without merging**